File: ts/src/scope-token.ts
// Serialisable attenuated scopes — the wire format behind `scope.toToken()` and
// `Watchlight.scopeFromToken()`.
//
// A scope token lets a scope derived in one process be re-established in
// another (a queue worker, a scheduler) WITHOUT trusting the job payload: the
// token is an HMAC over the canonical scope claims, and the receiving side
// re-runs the engine's strict-subset attenuation for every level of the chain
// before the scope is usable. The token is integrity, not authority — the
// engine, not the token, decides whether the requested chain is a subset.
//
// This module is pure (no engine, no I/O): canonicalisation, signing and
// verification. It is mirrored byte-for-byte by Python `watchlight.scope_token`.
//
// ── Token format ────────────────────────────────────────────────────────────
//
//   wls1.<base64url(canonical JSON claims)>.<base64url(HMAC-SHA256)>
//
//   * `wls1` is the format version; unknown versions are rejected.
//   * The HMAC key is the raw bytes of the token secret. The HMAC input is the
//     ASCII bytes of `"wls1." + base64url(payload)` — so the version is signed.
//   * base64url is RFC 4648 §5 with NO padding; non-canonical encodings and
//     padding are rejected.
//   * The whole token is bounded to MAX_TOKEN_LENGTH characters.
//
// ── Claims ──────────────────────────────────────────────────────────────────
//
//   {
//     "agent": "<governor identity>",
//     "chain": [ { "intents": [...], "resources": [...], "time_budget_seconds": N, "tools": [...] }, ... ],
//     "depth": <chain.length>,
//     "exp":   <epoch seconds>,
//     "iat":   <epoch seconds>,
//     "root":  { "intents": [...], "max_depth": N, "resources": [...], "time_budget_seconds": N, "tools": [...] }
//   }
//
//   `root` is the root scope's grant; `chain[i]` is the ENGINE-GRANTED scope at
//   depth i+1 (never a raw request), which the verifier replays as the request
//   for that level. Exactly these keys, no others.
//
// ── Canonical JSON (both lanes MUST agree byte-for-byte) ────────────────────
//
//   * Object keys sorted by Unicode code point; all keys are ASCII.
//   * No whitespace anywhere.
//   * `tools`, `resources`, `intents` are sets: sorted by Unicode code point
//     (NOT by UTF-16 code unit), duplicates preserved.
//   * Integers only for numbers (plain decimal, no exponent, no fraction);
//     every number is a non-negative safe integer.
//   * Strings escaped per RFC 8259 with the minimal escape set: `"` `\` and
//     control characters U+0000–U+001F (`\b \f \n \r \t`, else `\u00XX`);
//     everything else — including non-ASCII — is emitted literally (UTF-8).
//     Strings must be valid Unicode scalar values (no lone surrogates).

import { createHmac, timingSafeEqual } from "node:crypto";

/** Token format version prefix. */
export const SCOPE_TOKEN_PREFIX = "wls1";
/** Upper bound on a token's length (characters); longer input is rejected unparsed. */
export const MAX_TOKEN_LENGTH = 16_384;
/** Tolerated clock skew for a token whose `iat` is in the future (seconds). */
export const MAX_IAT_SKEW_SECONDS = 60;
/** Minimum token-secret length (bytes). */
export const MIN_SECRET_BYTES = 16;
/** Maximum chain length a token may carry (mirrors the DE depth ceiling). */
export const MAX_CHAIN_LENGTH = 5;

const HMAC_BYTES = 32;
const B64URL = /^[A-Za-z0-9_-]+$/;

/** The granted dimensions of one level of the chain. */
export interface ScopeStepClaim {
  tools: string[];
  resources: string[];
  intents: string[];
  time_budget_seconds: number;
}

/** The root scope's grant. */
export interface ScopeRootClaim extends ScopeStepClaim {
  max_depth: number;
}

/** Everything a scope token carries. */
export interface ScopeTokenClaims {
  agent: string;
  root: ScopeRootClaim;
  chain: ScopeStepClaim[];
  depth: number;
  iat: number;
  exp: number;
}

export type ScopeTokenErrorCode =
  | "no_secret"
  | "weak_secret"
  | "too_large"
  | "malformed"
  | "version"
  | "signature"
  | "identity"
  | "future_iat"
  | "expired"
  | "lifetime"
  | "mismatch";

/** Raised when a scope token cannot be minted or is rejected. The message never
 *  contains the token, its claims, or the secret. */
export class ScopeTokenError extends Error {
  readonly code: ScopeTokenErrorCode;
  constructor(code: ScopeTokenErrorCode, message: string) {
    super(`scope token rejected (${code}): ${message}`);
    this.name = "ScopeTokenError";
    this.code = code;
  }
}

// ── canonical JSON ──────────────────────────────────────────────────────────

/** Compare two strings by Unicode code point (Python `sorted()` order). */
export function codePointCompare(a: string, b: string): number {
  const ia = a[Symbol.iterator]();
  const ib = b[Symbol.iterator]();
  for (;;) {
    const na = ia.next();
    const nb = ib.next();
    if (na.done && nb.done) return 0;
    if (na.done) return -1;
    if (nb.done) return 1;
    const ca = na.value.codePointAt(0) as number;
    const cb = nb.value.codePointAt(0) as number;
    if (ca !== cb) return ca < cb ? -1 : 1;
  }
}

const isSafeNonNegInt = (n: unknown): n is number =>
  typeof n === "number" && Number.isSafeInteger(n) && n >= 0;

// A lone (unpaired) UTF-16 surrogate — not a Unicode scalar value.
const LONE_SURROGATE = /[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?<![\uD800-\uDBFF])[\uDC00-\uDFFF]/;

function canonicalString(s: string): string {
  if (LONE_SURROGATE.test(s)) throw new ScopeTokenError("malformed", "string is not valid Unicode");
  // JSON.stringify emits exactly the minimal escape set for well-formed strings.
  return JSON.stringify(s);
}

/** Serialise a claims value to canonical JSON. Arrays are emitted in the order
 *  given — callers sort set-valued arrays first (see {@link normalizeClaims}). */
export function canonicalJson(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) throw new ScopeTokenError("malformed", "numbers must be safe integers");
    return String(value);
  }
  if (typeof value === "string") return canonicalString(value);
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort(codePointCompare);
    return `{${keys.map((k) => `${canonicalString(k)}:${canonicalJson(obj[k])}`).join(",")}}`;
  }
  throw new ScopeTokenError("malformed", "unsupported value in claims");
}

const sortedSet = (xs: readonly string[]): string[] => [...xs].sort(codePointCompare);

function normalizeStep(s: ScopeStepClaim): ScopeStepClaim {
  return {
    tools: sortedSet(s.tools),
    resources: sortedSet(s.resources),
    intents: sortedSet(s.intents),
    time_budget_seconds: s.time_budget_seconds,
  };
}

/** Return a copy of `claims` with every set-valued array sorted — the exact
 *  object whose canonical JSON is signed. */
export function normalizeClaims(claims: ScopeTokenClaims): ScopeTokenClaims {
  return {
    agent: claims.agent,
    root: { ...normalizeStep(claims.root), max_depth: claims.root.max_depth },
    chain: claims.chain.map(normalizeStep),
    depth: claims.depth,
    iat: claims.iat,
    exp: claims.exp,
  };
}

// ── secret handling ─────────────────────────────────────────────────────────

/** Coerce a configured secret to bytes and enforce the minimum length. Never
 *  echoes the secret. */
export function normalizeSecret(secret: string | Uint8Array | undefined): Uint8Array | undefined {
  // An empty / whitespace-only value (an unfilled `.env` placeholder) is "unset":
  // constructing a governor must not fail for users who never mint a token —
  // the token operations themselves fail closed with `no_secret`.
  if (secret === undefined || (typeof secret === "string" && secret.trim().length === 0)) return undefined;
  if (typeof secret !== "string" && secret.length === 0) return undefined;
  // Copy caller-provided bytes so a later mutation of their array cannot change the key.
  const bytes = typeof secret === "string" ? Buffer.from(secret, "utf8") : Buffer.from(secret);
  if (bytes.length < MIN_SECRET_BYTES) {
    throw new ScopeTokenError("weak_secret", `token secret must be at least ${MIN_SECRET_BYTES} bytes`);
  }
  return bytes;
}

export function requireSecret(secret: Uint8Array | undefined): Uint8Array {
  if (!secret) {
    throw new ScopeTokenError(
      "no_secret",
      "scope tokens require a token secret — construct Watchlight with { tokenSecret } " +
        "(or set WATCHLIGHT_TOKEN_SECRET); there is no default"
    );
  }
  return secret;
}

// ── sign / verify ───────────────────────────────────────────────────────────

const b64url = (bytes: Uint8Array): string => Buffer.from(bytes).toString("base64url");

/** Strict base64url decode: alphabet-only, no padding, canonical encoding. */
function b64urlDecode(s: string): Uint8Array {
  if (s.length === 0 || !B64URL.test(s) || s.length % 4 === 1) {
    throw new ScopeTokenError("malformed", "token segment is not base64url");
  }
  const bytes = Buffer.from(s, "base64url");
  if (bytes.toString("base64url") !== s) {
    throw new ScopeTokenError("malformed", "token segment is not canonical base64url");
  }
  return bytes;
}

function hmac(secret: Uint8Array, signedPart: string): Buffer {
  return createHmac("sha256", secret).update(signedPart, "utf8").digest();
}

/** Mint a token over `claims` (normalised + canonicalised here). */
export function signScopeToken(claims: ScopeTokenClaims, secret: Uint8Array): string {
  const payload = Buffer.from(canonicalJson(normalizeClaims(claims)), "utf8");
  const signedPart = `${SCOPE_TOKEN_PREFIX}.${b64url(payload)}`;
  const token = `${signedPart}.${b64url(hmac(secret, signedPart))}`;
  if (token.length > MAX_TOKEN_LENGTH) {
    throw new ScopeTokenError("too_large", `token exceeds ${MAX_TOKEN_LENGTH} characters`);
  }
  return token;
}

function expectStringArray(v: unknown, what: string): string[] {
  if (!Array.isArray(v) || !v.every((x) => typeof x === "string")) {
    throw new ScopeTokenError("malformed", `${what} must be a list of strings`);
  }
  return v as string[];
}

function expectInt(v: unknown, what: string): number {
  if (!isSafeNonNegInt(v)) throw new ScopeTokenError("malformed", `${what} must be a non-negative integer`);
  return v;
}

function expectExactKeys(obj: unknown, keys: readonly string[], what: string): Record<string, unknown> {
  if (obj === null || typeof obj !== "object" || Array.isArray(obj)) {
    throw new ScopeTokenError("malformed", `${what} must be an object`);
  }
  const rec = obj as Record<string, unknown>;
  const have = Object.keys(rec).sort(codePointCompare);
  const want = [...keys].sort(codePointCompare);
  if (have.length !== want.length || have.some((k, i) => k !== want[i])) {
    throw new ScopeTokenError("malformed", `${what} has unexpected or missing fields`);
  }
  return rec;
}

const STEP_KEYS = ["intents", "resources", "time_budget_seconds", "tools"] as const;
const ROOT_KEYS = [...STEP_KEYS, "max_depth"] as const;
const CLAIM_KEYS = ["agent", "chain", "depth", "exp", "iat", "root"] as const;

function parseStep(v: unknown, what: string): ScopeStepClaim {
  const o = expectExactKeys(v, STEP_KEYS, what);
  return {
    tools: expectStringArray(o.tools, `${what}.tools`),
    resources: expectStringArray(o.resources, `${what}.resources`),
    intents: expectStringArray(o.intents, `${what}.intents`),
    time_budget_seconds: expectInt(o.time_budget_seconds, `${what}.time_budget_seconds`),
  };
}

/** Validate an already-authenticated payload against the exact claims schema. */
export function parseClaims(raw: unknown): ScopeTokenClaims {
  const o = expectExactKeys(raw, CLAIM_KEYS, "claims");
  if (typeof o.agent !== "string" || o.agent.length === 0) {
    throw new ScopeTokenError("malformed", "claims.agent must be a non-empty string");
  }
  const rootRec = expectExactKeys(o.root, ROOT_KEYS, "claims.root");
  const root: ScopeRootClaim = {
    ...parseStep(
      { tools: rootRec.tools, resources: rootRec.resources, intents: rootRec.intents, time_budget_seconds: rootRec.time_budget_seconds },
      "claims.root"
    ),
    max_depth: expectInt(rootRec.max_depth, "claims.root.max_depth"),
  };
  if (root.max_depth > MAX_CHAIN_LENGTH) {
    throw new ScopeTokenError("malformed", `claims.root.max_depth must not exceed ${MAX_CHAIN_LENGTH}`);
  }
  if (!Array.isArray(o.chain) || o.chain.length > MAX_CHAIN_LENGTH) {
    throw new ScopeTokenError("malformed", `claims.chain must be a list of at most ${MAX_CHAIN_LENGTH} steps`);
  }
  const chain = o.chain.map((s, i) => parseStep(s, `claims.chain[${i}]`));
  const depth = expectInt(o.depth, "claims.depth");
  if (depth !== chain.length) throw new ScopeTokenError("malformed", "claims.depth must equal the chain length");
  const iat = expectInt(o.iat, "claims.iat");
  const exp = expectInt(o.exp, "claims.exp");
  return { agent: o.agent, root, chain, depth, iat, exp };
}

/** Whole-second epoch clock. */
export const nowSeconds = (): number => Math.floor(Date.now() / 1000);

/**
 * Verify a token's shape, signature (constant-time), identity binding and time
 * window, and return its claims. Does NOT touch the engine — the caller must
 * still replay the chain through `attenuate()` before trusting the scope.
 */
export function verifyScopeToken(
  token: unknown,
  secret: Uint8Array,
  opts: { agent: string; now?: number }
): ScopeTokenClaims {
  if (typeof token !== "string") throw new ScopeTokenError("malformed", "token must be a string");
  if (token.length > MAX_TOKEN_LENGTH) {
    throw new ScopeTokenError("too_large", `token exceeds ${MAX_TOKEN_LENGTH} characters`);
  }
  const parts = token.split(".");
  if (parts.length !== 3) throw new ScopeTokenError("malformed", "token must have three segments");
  const [version, payloadB64, sigB64] = parts;
  if (version !== SCOPE_TOKEN_PREFIX) throw new ScopeTokenError("version", "unsupported token version");

  // Shape checks only (strict base64url; decoding is not parsing), then
  // authenticate BEFORE anything inside the payload is interpreted.
  const payload = b64urlDecode(payloadB64);
  const provided = b64urlDecode(sigB64);
  const expected = hmac(secret, `${version}.${payloadB64}`);
  if (provided.length !== HMAC_BYTES || !timingSafeEqual(provided, expected)) {
    throw new ScopeTokenError("signature", "signature does not verify");
  }

  let raw: unknown;
  try {
    raw = JSON.parse(Buffer.from(payload).toString("utf8"));
  } catch {
    throw new ScopeTokenError("malformed", "payload is not JSON");
  }
  const claims = parseClaims(raw);
  // The payload must be in canonical form — a re-encoding of the parsed claims
  // that differs from what was signed is rejected (no alternate encodings).
  if (canonicalJson(claims) !== Buffer.from(payload).toString("utf8")) {
    throw new ScopeTokenError("malformed", "payload is not canonical");
  }

  if (claims.agent !== opts.agent) throw new ScopeTokenError("identity", "token is bound to a different agent");

  const now = opts.now ?? nowSeconds();
  if (claims.iat > now + MAX_IAT_SKEW_SECONDS) throw new ScopeTokenError("future_iat", "token issued in the future");
  if (claims.exp <= now) throw new ScopeTokenError("expired", "token has expired");
  if (claims.exp <= claims.iat) throw new ScopeTokenError("lifetime", "token expires before it was issued");
  // A token may never outlive the scope it names: its lifetime is bounded by
  // the (engine-clamped) time budget of the deepest level it carries.
  const budget = claims.chain.length ? claims.chain[claims.chain.length - 1].time_budget_seconds : claims.root.time_budget_seconds;
  if (claims.exp - claims.iat > budget) {
    throw new ScopeTokenError("lifetime", "token lifetime exceeds the scope's time budget");
  }
  return claims;
}

/** Set-equality on string lists (order-insensitive, multiplicity-sensitive). */
export function sameSet(a: readonly string[], b: readonly string[]): boolean {
  if (a.length !== b.length) return false;
  const sa = sortedSet(a);
  const sb = sortedSet(b);
  return sa.every((x, i) => x === sb[i]);
}

File: src/watchlight/scope_token.py
"""Serialisable attenuated scopes — the wire format behind ``Scope.to_token()``
and ``Watchlight.scope_from_token()``.

A scope token lets a scope derived in one process be re-established in another
(a queue worker, a scheduler) WITHOUT trusting the job payload: the token is an
HMAC over the canonical scope claims, and the receiving side re-runs the
engine's strict-subset attenuation for every level of the chain before the scope
is usable. The token is integrity, not authority — the engine, not the token,
decides whether the requested chain is a subset.

This module is pure (no engine, no I/O): canonicalisation, signing and
verification. It mirrors ``ts/src/scope-token.ts`` byte-for-byte.

Token format::

    wls1.<base64url(canonical JSON claims)>.<base64url(HMAC-SHA256)>

* ``wls1`` is the format version; unknown versions are rejected.
* The HMAC key is the raw bytes of the token secret. The HMAC input is the ASCII
  bytes of ``"wls1." + base64url(payload)`` — so the version is signed.
* base64url is RFC 4648 §5 with NO padding; non-canonical encodings and padding
  are rejected.
* The whole token is bounded to :data:`MAX_TOKEN_LENGTH` characters.

Claims::

    {
      "agent": "<governor identity>",
      "chain": [ {"intents": [...], "resources": [...], "time_budget_seconds": N, "tools": [...]}, ... ],
      "depth": <len(chain)>,
      "exp":   <epoch seconds>,
      "iat":   <epoch seconds>,
      "root":  {"intents": [...], "max_depth": N, "resources": [...], "time_budget_seconds": N, "tools": [...]}
    }

``root`` is the root scope's grant; ``chain[i]`` is the ENGINE-GRANTED scope at
depth ``i+1`` (never a raw request), which the verifier replays as the request
for that level. Exactly these keys, no others.

Canonical JSON (both lanes MUST agree byte-for-byte):

* Object keys sorted by Unicode code point; all keys are ASCII.
* No whitespace anywhere.
* ``tools``, ``resources``, ``intents`` are sets: sorted by Unicode code point,
  duplicates preserved.
* Integers only for numbers (plain decimal); every number is a non-negative
  integer below 2**53.
* Strings escaped per RFC 8259 with the minimal escape set: ``"`` ``\\`` and
  control characters U+0000–U+001F (``\\b \\f \\n \\r \\t``, else ``\\u00XX``);
  everything else — including non-ASCII — is emitted literally (UTF-8). Strings
  must be valid Unicode scalar values (no lone surrogates).
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import re
import time
from typing import Any, Optional, Sequence, Union

__all__ = [
    "ScopeTokenError",
    "SCOPE_TOKEN_PREFIX",
    "MAX_TOKEN_LENGTH",
    "MAX_IAT_SKEW_SECONDS",
    "MIN_SECRET_BYTES",
    "MAX_CHAIN_LENGTH",
    "canonical_json",
    "normalize_claims",
    "sign_scope_token",
    "verify_scope_token",
]

#: Token format version prefix.
SCOPE_TOKEN_PREFIX = "wls1"
#: Upper bound on a token's length (characters); longer input is rejected unparsed.
MAX_TOKEN_LENGTH = 16_384
#: Tolerated clock skew for a token whose ``iat`` is in the future (seconds).
MAX_IAT_SKEW_SECONDS = 60
#: Minimum token-secret length (bytes).
MIN_SECRET_BYTES = 16
#: Maximum chain length a token may carry (mirrors the DE depth ceiling).
MAX_CHAIN_LENGTH = 5

_HMAC_BYTES = 32
_MAX_SAFE_INT = 2**53 - 1
_B64URL = re.compile(r"^[A-Za-z0-9_-]+$")

_STEP_KEYS = ("intents", "resources", "time_budget_seconds", "tools")
_ROOT_KEYS = _STEP_KEYS + ("max_depth",)
_CLAIM_KEYS = ("agent", "chain", "depth", "exp", "iat", "root")


class ScopeTokenError(PermissionError):
    """Raised when a scope token cannot be minted or is rejected. The message
    never contains the token, its claims, or the secret."""

    def __init__(self, code: str, message: str) -> None:
        self.code = code
        super().__init__(f"scope token rejected ({code}): {message}")


# ── canonical JSON ──────────────────────────────────────────────────────────


def _canonical_string(s: str) -> str:
    try:
        s.encode("utf-8")
    except UnicodeEncodeError:
        raise ScopeTokenError("malformed", "string is not valid Unicode") from None
    # ensure_ascii=False emits exactly the minimal RFC 8259 escape set.
    return json.dumps(s, ensure_ascii=False)


def canonical_json(value: Any) -> str:
    """Serialise a claims value to canonical JSON. Lists are emitted in the order
    given — callers sort set-valued lists first (see :func:`normalize_claims`)."""
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, int):
        if not 0 <= value <= _MAX_SAFE_INT:
            raise ScopeTokenError("malformed", "numbers must be non-negative integers below 2**53")
        return str(value)
    if isinstance(value, str):
        return _canonical_string(value)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonical_json(v) for v in value) + "]"
    if isinstance(value, dict):
        # Python sorts str by code point — the same order the TS lane uses.
        keys = sorted(value)
        return "{" + ",".join(f"{_canonical_string(k)}:{canonical_json(value[k])}" for k in keys) + "}"
    raise ScopeTokenError("malformed", "unsupported value in claims")


def _sorted_set(xs: Sequence[str]) -> list[str]:
    return sorted(xs)


def _normalize_step(step: dict[str, Any]) -> dict[str, Any]:
    return {
        "tools": _sorted_set(step["tools"]),
        "resources": _sorted_set(step["resources"]),
        "intents": _sorted_set(step["intents"]),
        "time_budget_seconds": step["time_budget_seconds"],
    }


def normalize_claims(claims: dict[str, Any]) -> dict[str, Any]:
    """Return a copy of ``claims`` with every set-valued list sorted — the exact
    object whose canonical JSON is signed."""
    root = _normalize_step(claims["root"])
    root["max_depth"] = claims["root"]["max_depth"]
    return {
        "agent": claims["agent"],
        "root": root,
        "chain": [_normalize_step(s) for s in claims["chain"]],
        "depth": claims["depth"],
        "iat": claims["iat"],
        "exp": claims["exp"],
    }


# ── secret handling ─────────────────────────────────────────────────────────


def normalize_secret(secret: Union[str, bytes, bytearray, None]) -> Optional[bytes]:
    """Coerce a configured secret to bytes and enforce the minimum length. Never
    echoes the secret."""
    if secret is None:
        return None
    raw = secret.encode("utf-8") if isinstance(secret, str) else bytes(secret)
    if len(raw) < MIN_SECRET_BYTES:
        raise ScopeTokenError("weak_secret", f"token secret must be at least {MIN_SECRET_BYTES} bytes")
    return raw


def require_secret(secret: Optional[bytes]) -> bytes:
    if not secret:
        raise ScopeTokenError(
            "no_secret",
            "scope tokens require a token secret — construct Watchlight(token_secret=...) "
            "(or set WATCHLIGHT_TOKEN_SECRET); there is no default",
        )
    return secret


# ── sign / verify ───────────────────────────────────────────────────────────


def _b64url(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def _b64url_decode(s: str) -> bytes:
    """Strict base64url decode: alphabet-only, no padding, canonical encoding."""
    if not s or not _B64URL.match(s) or len(s) % 4 == 1:
        raise ScopeTokenError("malformed", "token segment is not base64url")
    padded = s + "=" * (-len(s) % 4)
    try:
        raw = base64.b64decode(padded.replace("-", "+").replace("_", "/"), validate=True)
    except (ValueError, TypeError):
        raise ScopeTokenError("malformed", "token segment is not base64url") from None
    if _b64url(raw) != s:
        raise ScopeTokenError("malformed", "token segment is not canonical base64url")
    return raw


def _hmac(secret: bytes, signed_part: str) -> bytes:
    return hmac.new(secret, signed_part.encode("utf-8"), hashlib.sha256).digest()


def sign_scope_token(claims: dict[str, Any], secret: bytes) -> str:
    """Mint a token over ``claims`` (normalised + canonicalised here)."""
    payload = canonical_json(normalize_claims(claims)).encode("utf-8")
    signed_part = f"{SCOPE_TOKEN_PREFIX}.{_b64url(payload)}"
    token = f"{signed_part}.{_b64url(_hmac(secret, signed_part))}"
    if len(token) > MAX_TOKEN_LENGTH:
        raise ScopeTokenError("too_large", f"token exceeds {MAX_TOKEN_LENGTH} characters")
    return token


def _expect_string_list(v: Any, what: str) -> list[str]:
    if not isinstance(v, list) or not all(isinstance(x, str) for x in v):
        raise ScopeTokenError("malformed", f"{what} must be a list of strings")
    return v


def _expect_int(v: Any, what: str) -> int:
    if isinstance(v, bool) or not isinstance(v, int) or not 0 <= v <= _MAX_SAFE_INT:
        raise ScopeTokenError("malformed", f"{what} must be a non-negative integer")
    return v


def _expect_exact_keys(obj: Any, keys: Sequence[str], what: str) -> dict[str, Any]:
    if not isinstance(obj, dict):
        raise ScopeTokenError("malformed", f"{what} must be an object")
    if sorted(obj) != sorted(keys):
        raise ScopeTokenError("malformed", f"{what} has unexpected or missing fields")
    return obj


def _parse_step(v: Any, what: str) -> dict[str, Any]:
    o = _expect_exact_keys(v, _STEP_KEYS, what)
    return {
        "tools": _expect_string_list(o["tools"], f"{what}.tools"),
        "resources": _expect_string_list(o["resources"], f"{what}.resources"),
        "intents": _expect_string_list(o["intents"], f"{what}.intents"),
        "time_budget_seconds": _expect_int(o["time_budget_seconds"], f"{what}.time_budget_seconds"),
    }


def parse_claims(raw: Any) -> dict[str, Any]:
    """Validate an already-authenticated payload against the exact claims schema."""
    o = _expect_exact_keys(raw, _CLAIM_KEYS, "claims")
    agent = o["agent"]
    if not isinstance(agent, str) or not agent:
        raise ScopeTokenError("malformed", "claims.agent must be a non-empty string")
    root_rec = _expect_exact_keys(o["root"], _ROOT_KEYS, "claims.root")
    root = _parse_step({k: root_rec[k] for k in _STEP_KEYS}, "claims.root")
    root["max_depth"] = _expect_int(root_rec["max_depth"], "claims.root.max_depth")
    if root["max_depth"] > MAX_CHAIN_LENGTH:
        raise ScopeTokenError("malformed", f"claims.root.max_depth must not exceed {MAX_CHAIN_LENGTH}")
    chain_raw = o["chain"]
    if not isinstance(chain_raw, list) or len(chain_raw) > MAX_CHAIN_LENGTH:
        raise ScopeTokenError("malformed", f"claims.chain must be a list of at most {MAX_CHAIN_LENGTH} steps")
    chain = [_parse_step(s, f"claims.chain[{i}]") for i, s in enumerate(chain_raw)]
    depth = _expect_int(o["depth"], "claims.depth")
    if depth != len(chain):
        raise ScopeTokenError("malformed", "claims.depth must equal the chain length")
    iat = _expect_int(o["iat"], "claims.iat")
    exp = _expect_int(o["exp"], "claims.exp")
    return {"agent": agent, "root": root, "chain": chain, "depth": depth, "iat": iat, "exp": exp}


def now_seconds() -> int:
    """Whole-second epoch clock."""
    return int(time.time())


def verify_scope_token(token: Any, secret: bytes, *, agent: str, now: Optional[int] = None) -> dict[str, Any]:
    """Verify a token's shape, signature (constant-time), identity binding and
    time window, and return its claims. Does NOT touch the engine — the caller
    must still replay the chain through ``attenuate()`` before trusting the scope."""
    if not isinstance(token, str):
        raise ScopeTokenError("malformed", "token must be a string")
    if len(token) > MAX_TOKEN_LENGTH:
        raise ScopeTokenError("too_large", f"token exceeds {MAX_TOKEN_LENGTH} characters")
    parts = token.split(".")
    if len(parts) != 3:
        raise ScopeTokenError("malformed", "token must have three segments")
    version, payload_b64, sig_b64 = parts
    if version != SCOPE_TOKEN_PREFIX:
        raise ScopeTokenError("version", "unsupported token version")

    # Shape checks only (strict base64url; decoding is not parsing), then
    # authenticate BEFORE anything inside the payload is interpreted.
    payload = _b64url_decode(payload_b64)
    provided = _b64url_decode(sig_b64)
    expected = _hmac(secret, f"{version}.{payload_b64}")
    if len(provided) != _HMAC_BYTES or not hmac.compare_digest(provided, expected):
        raise ScopeTokenError("signature", "signature does not verify")

    try:
        raw = json.loads(payload.decode("utf-8"))
    except (ValueError, UnicodeDecodeError):
        raise ScopeTokenError("malformed", "payload is not JSON") from None
    claims = parse_claims(raw)
    # The payload must be in canonical form — a re-encoding of the parsed claims
    # that differs from what was signed is rejected (no alternate encodings).
    if canonical_json(claims).encode("utf-8") != payload:
        raise ScopeTokenError("malformed", "payload is not canonical")

    if claims["agent"] != agent:
        raise ScopeTokenError("identity", "token is bound to a different agent")

    now_s = now_seconds() if now is None else now
    if claims["iat"] > now_s + MAX_IAT_SKEW_SECONDS:
        raise ScopeTokenError("future_iat", "token issued in the future")
    if claims["exp"] <= now_s:
        raise ScopeTokenError("expired", "token has expired")
    if claims["exp"] <= claims["iat"]:
        raise ScopeTokenError("lifetime", "token expires before it was issued")
    # A token may never outlive the scope it names: its lifetime is bounded by
    # the (engine-clamped) time budget of the deepest level it carries.
    budget = (claims["chain"][-1] if claims["chain"] else claims["root"])["time_budget_seconds"]
    if claims["exp"] - claims["iat"] > budget:
        raise ScopeTokenError("lifetime", "token lifetime exceeds the scope's time budget")
    return claims


def same_set(a: Sequence[str], b: Sequence[str]) -> bool:
    """Set-equality on string lists (order-insensitive, multiplicity-sensitive)."""
    return len(a) == len(b) and sorted(a) == sorted(b)
